tests/test_packages.py

```diff
@@ -157,7 +141,12 @@ def test_package__both_mappings():
             """
             ],
             None,
-            {"apache_airflow": ["airflow"], "attrs": ["attr", "attrs"]},
+            {
+                "apache_airflow": Package(
+                    "apache_airflow", {"airflow"}, UserDefinedMapping
```

Comment:
There is one consideration, that we may move to a separate issue. Right now it is not clear if a `package_name` field in a `Package` object should be normalized at the creation of the object or not.

I do not see any case now when unnormalized name may be of use. We could simplify this logic and add a check fo `Package` object that the package name is normalized. Do you recall any case where unnormalized name is needed?

Reply:
TL;DR: I believe you are right, and we can indeed make name normalization mandatory when creating `Package` objects. But I'd rather wait until later in these PRs. Digression/rationale below:

---

Thinking more about non-normalized package names:

After this PR and #321, I believe `IdentityMapping` is the only resolver where the `Package` object still keeps the original/non-normalized package name, but even there it is of no practical use.

The non-normalized package names do still exist as _keys_ in the dictionary returned from `resolve_dependencies()` (and `BasePackageResolver.lookup_packages()`). This is still correct/necessary, as we use the non-normalized package names from `DeclaredDependency` to do lookups into this dict.

Sidebar: There are actually _two different_ sources of non-normalized package names:

- The dependency declarations returned from `extract_declared_dependencies`. These are preserved as the keys in our `resolve_dependencies()` dict as explained above.
- The `dist.name` member returned from `importlib_metadata.MetadataPathFinder.find_distributions()`, which is ultimately parsed from the installed package's metadata. These were originally preserved (but never really used) in the `.package_name` member of `Package` objects returned from `LocalPackageResolver`. But after the first commit in #321 (where we teach `LocalPackageResolver` to merge `Package` objects from different Python environments with `accumulate_mappings()`) these are finally gone.